### PR TITLE
Initial rough dialback support

### PR DESCRIPTION
**Pull request by jhass**
* [x] No tests
* [x] Enable tls
* [x] ~~No cluster support (Only if upstream interested?)~~ see #52 
* [x] ~~Investigate also adding Bidi (XEP-0288) support  (Only if upstream interested?)~~ see #53

### CODE REVIEW

File: lib/vines/stanza/dialback.rb [hunk 1]
```diff
@@ -0,0 +1,30 @@
+# encoding: UTF-8
+
+module Vines
+  class Stanza
+    class Dialback < Stanza
+      VALID_TYPE, INVALID_TYPE = %w[valid invalid].map {|t| t.freeze }
+      NS = NAMESPACES[:legacy_dialback]
+
+      register "/db:verify", 'db' => NS
```
Comment (Zauberstuhl) [inline lib/vines/stanza/dialback.rb]: ok where do you listen for `db:result`?

* Initiating Server Sends Dialback Key (Step 1) -> `db:result`
* Receiving Server Sends Verification Request to Authoritative Server (Step 2) `db:verify`

`db:result` ???? `db:verify` -> `db:verify (In)Valid` -> `db:result (In)Valid`

Comment (jhass) [inline lib/vines/stanza/dialback.rb]: > ok where do you listen for db:result?

Scroll down, I pointed that out in the two comments I made after this one.

File: lib/vines/stream/server/outbound/auth.rb [hunk 2]
```diff
@@ -5,24 +5,54 @@ class Stream
...
+            authzid = Base64.strict_encode64(stream.domain)
+            stream.write(%Q{<auth xmlns="#{NAMESPACES[:sasl]}" mechanism="EXTERNAL">#{authzid}</auth>})
+            advance
+          end
+
+          def initiate_dialback!
+            secret = Kit.auth_token
+            dialback_key = Kit.dialback_key(secret, stream.remote_domain, stream.domain, stream.id)
+
+            stream.write(%Q(<db:result from="#{stream.domain}" to="#{stream.remote_domain}">#{dialback_key}</db:result>))
```
Comment (jhass) [inline lib/vines/stream/server/outbound/auth.rb]: This is where the initiating server initiates the dialback procedure, since it isn't offered SASL authentication because the server it connects to couldn't validate its certificate on one of its outbound connections. Is that the part you're missing maybe?

**Comment by Zauberstuhl** [comment]
plain dialback works now

last two remaining points:

* enable tls (external)
* write tests

File: lib/vines/stanza/presence.rb [hunk 3]
```diff
@@ -18,7 +18,7 @@ def process
         unless self['type'].nil?
           raise StanzaErrors::BadRequest.new(self, 'modify')
         end
-        if Config.instance.max_offline_msgs > 0
+        if Config.instance.max_offline_msgs > 0 && !validate_to
```
Comment (jhass) [inline lib/vines/stanza/presence.rb]: Seems unrelated to the PR?

Comment (Zauberstuhl) [inline lib/vines/stanza/presence.rb]: right! I'll separate it

**Comment by Zauberstuhl** [comment]
Rebased against develop and squashed test-commits.
Functionality is fully implemented. We should start a code review!

File: lib/vines/stream.rb [hunk 4]
```diff
@@ -303,5 +303,13 @@ def state
     def valid_address?(jid)
       JID.new(jid).domain? rescue false
     end
+
+    def update_stream_id(id_or_node)
```
Comment (Zauberstuhl) [inline lib/vines/stream.rb]: have to check why we needed the stream id so early .. ?

File: lib/vines/config/host.rb [hunk 5]
```diff
@@ -30,6 +31,14 @@ def storage(name=nil, &block)
         end
       end
 
+      def force_s2s_encryption(enabled)
+        @force_s2s_encryption = !!enabled
```
Comment (jhass) [inline lib/vines/config/host.rb]: Which values that are not `true` or `false` do you expect here?

File: lib/vines/stream.rb [hunk 6]
```diff
@@ -201,6 +202,12 @@ def router
       @config.router
     end
 
+    # Returns the current +State+ of the stream's state machine. Provided as a
+    # method so subclasses can override the behavior.
+    def state
+      @state
+    end
```
Comment (jhass) [inline lib/vines/stream.rb]: Actually this should be a `attr_reader` now.

File: lib/vines/stream/server.rb [hunk 7]
```diff
@@ -72,8 +75,27 @@ def max_stanza_size
         config[:server].max_stanza_size
       end
 
+      def ssl_verify_peer(pem)
+        @store.trusted?(pem).tap {|trusted| @peer_trusted = trusted}
```
Comment (Zauberstuhl) [inline lib/vines/stream/server.rb]: that overrides the stream class function. we should be able to use the skip flag here

File: lib/vines/stream/server.rb [hunk 8]
```diff
@@ -15,7 +15,7 @@ class Server < Stream
       # yielded stream will be nil if the remote connection failed. We need to
       # use a background thread to avoid blocking the server on DNS SRV
       # lookups.
-      def self.start(config, to, from, &callback)
+      def self.start(config, to, from, dbv = false, &callback)
```
Comment (jhass) [inline lib/vines/stream/server.rb]: `dbv` is not a good name, don't be afraid of becoming too verbose.

File: lib/vines/stream/server.rb [hunk 9]
```diff
@@ -53,10 +53,13 @@ def self.connect(config, to, from, srv, callback)
 
       def initialize(config, options={})
         super(config)
+        @outbound_tls_required = false
+        @peer_trusted = nil
         @connected = false
         @remote_domain = options[:to]
         @domain = options[:from]
         @srv = options[:srv]
+        @dialback_verify = options[:dialback_verify]
```
Comment (jhass) [inline lib/vines/stream/server.rb]: `dialback_verify` what? `in_dialback_verification`? `dialback_verify_key`?

File: lib/vines/stream/server.rb (hunk 7 above)
Comment (jhass) [inline lib/vines/stream/server.rb]: The tap is twice unnecessary here. For one, the value of an assignment expression is always the right hand side and for the second we don't even use this expression as return value of the method.

File: lib/vines/stream/server.rb [hunk 10]
```diff
@@ -72,8 +75,27 @@ def max_stanza_size
...
       end
 
+      def ssl_verify_peer(pem)
+        @store.trusted?(pem).tap {|trusted| @peer_trusted = trusted}
+        true
+      end
+
       def ssl_handshake_completed
-        close_connection unless cert_domain_matches?(@remote_domain)
+        @peer_trusted = cert_domain_matches?(@remote_domain) && @peer_trusted
```
Comment (jhass) [inline lib/vines/stream/server.rb]: Nitpick, but maybe add a `peer_trusted?` method so the level of abstraction is kept the same in this method.

File: lib/vines/stream/server.rb [hunk 11]
```diff
@@ -72,8 +75,27 @@ def max_stanza_size
...
+      end
+
+      def dialback_retry?
+        !@peer_trusted.nil? && !@peer_trusted
```
Comment (jhass) [inline lib/vines/stream/server.rb]: I find this a bit hard to follow, I'd say do:
```ruby
return false if @peer_trusted.nil? || @peer_trusted
true
```

a bit more verbose, but takes out all those negations you need to think about.

File: lib/vines/stream/server.rb [hunk 12]
```diff
@@ -72,8 +75,27 @@ def max_stanza_size
...
+      end
+
+      def outbound_tls_required?
+        @outbound_tls_required
+      end
+
+      def outbound_tls_required(required)
+        if !!required == required
```
Comment (jhass) [inline lib/vines/stream/server.rb]: This seems strange, can you elaborate?

File: lib/vines/stream/server/auth_method.rb [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+              stream.reset
+              advance
+            else
+              stream.write(FAILURE)
+              stream.write('</stream:stream>')
+              stream.close_connection_after_writing
+            end
+          elsif dialback?(node)
+            begin
+              Vines::Stream::Server.start(stream.config, node[FROM], node[TO], dbv = true) do |a|
```
Comment (jhass) [inline lib/vines/stream/server/auth_method.rb]: `dbv = true` just assigns a useless local variable, it's not a keyword argument.

File: lib/vines/stream/server/auth_method.rb [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+                a.write("<db:verify from='#{node[TO]}' id='#{stream.id}' to='#{node[FROM]}'>#{node.text}</db:verify>") if a
```
Comment (jhass) [inline lib/vines/stream/server/auth_method.rb]: Let's not hide the if at the end of a long line, use a regular if here.

File: lib/vines/stream/server/outbound/auth.rb [hunk 15]
```diff
@@ -5,24 +5,53 @@ class Stream
...
           end
 
           private
 
-          def external?(node)
-            external = node.xpath("ns:mechanisms/ns:mechanism[text()='EXTERNAL']", 'ns' => NS).any?
-            node.name == 'features' && namespace(node) == NAMESPACES[:stream] && external
+          def tls_required?(node)
+            child = node.xpath('ns:starttls', 'ns' => NAMESPACES[:tls]).children.first
+            !child.nil? && child.name == REQUIRED
```
Comment (jhass) [inline lib/vines/stream/server/outbound/auth.rb]: `nil` is falsey, just `child &&`

File: lib/vines/stream/server/outbound/authoritative.rb [hunk 16]
```diff
@@ -0,0 +1,48 @@
...
+          def initialize(stream, success=nil)
+            super
+          end
+
+          def node(node)
+            raise StreamErrors::NotAuthorized unless authoritative?(node)
+
+            case node[TYPE]
+            when VALID
+              @inbound.write("<db:result xmlns:db='#{NAMESPACES[:legacy_dialback]}' " +
```
Comment (jhass) [inline lib/vines/stream/server/outbound/authoritative.rb]: String#+ creates useless strings in most cases, as it returns a new string instead of modifying, so while constructing String#<< is to prefer.

The second way to write this over multiple lines is with a percent literal: `%(I'm a string)`, however that might include unwanted whitespace from the indentation.

So there's a third way commonly seen, utilizing the fact that `"foo" "bar"` is perfectly valid and equivalent to `"foobar"`, it's parsed as a single string literal. That combined with the line continuation gives a fairly pleasant result while still avoiding the rather verbose heredoc syntax:

```ruby
puts  "foo" \
         "bar"
```

File: lib/vines/stream/server/start.rb [hunk 17]
```diff
@@ -3,10 +3,35 @@
...
+
+        def node(node)
+          raise StreamErrors::NotAuthorized unless stream?(node)
+          stream.start(node)
+          doc = Document.new
+          features = doc.create_element('stream:features') do |el|
+            el << doc.create_element('starttls') do |tls|
+              tls.default_namespace = NAMESPACES[:tls]
+              tls << doc.create_element('required') if force_s2s_encryption?
+            end unless stream.dialback_retry?
```
Comment (jhass) [inline lib/vines/stream/server/start.rb]: I would refrain from the modifer-unless here.

**Comment by jhass** [comment]
So, to be honest I kept it to style review and didn't do a functionality review, I hope that's enough for you :)

**Comment by Zauberstuhl** [comment]
Thanks a lot! I will clean up!

Related to functional tests; I will take care of it, that's for sure! Since we have no QA (which I can blame)
I have to test my code on my own :P
